File: app/src/main/java/com/github/panda_io/micro_panda/ast/type/Function.java
package com.github.panda_io.micro_panda.ast.type;

import java.util.ArrayList;

public class Function extends Type {
    public Type returnType;
    public ArrayList<Type> parameters;

    public boolean isMemberFunction;
    public boolean isExtern;
    public String externName;
    public boolean isDefine;

    public Function() {
        this.parameters = new ArrayList<>();
    }

    public boolean equal(Type type) {
        if (type instanceof Function) {
            Function function = (Function) type;
            if (this.returnType != null && function.returnType != null) {
                if (!this.returnType.equal(function.returnType)) {
                    return false;
                }
            } else if (this.returnType != null || function.returnType != null) {
                return false;
            }
            if (this.parameters.size() != function.parameters.size()) {
                return false;
            }
            for (int i = 0; i < this.parameters.size(); i++) {
                if (!this.parameters.get(i).equal(function.parameters.get(i))) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    public String string() {
        StringBuffer buffer = new StringBuffer();
        buffer.append("function(");
        for (int i = 0; i < this.parameters.size(); i++) {
            if (i != 0) {
                buffer.append(", ");
            }
            buffer.append(this.parameters.get(i).string());
        }
        return buffer.toString();
    }
}

File: MicroPanda/AST/Type/Function.cs
namespace MicroPanda.AST.Type;



internal class TypeFunction : Type
{
    internal Type? ReturnType { get; set; }
    internal List<Type>? Parameters { get; set; }

    internal bool MemberFunction { get; set; }
    internal bool Extern { get; set; }
    internal string? ExternName { get; set; }
    internal bool TypeDefine { get; set; }

    override internal bool Equal(Type type)
    {
        if (type is TypeFunction typeFunction)
        {
            if (ReturnType != null && typeFunction.ReturnType != null)
            {
                if (!ReturnType.Equal(typeFunction.ReturnType))
                {
                    return false;
                }
            }
            else if (ReturnType != typeFunction.ReturnType)
            {
                return false;
            }

            if (Parameters!.Count != typeFunction.Parameters!.Count)
            {
                return false;
            }

            for (int i = 0; i < Parameters.Count; i++)
            {
                if (!Parameters[i].Equal(typeFunction.Parameters[i]))
                {
                    return false;
                }
            }

            return true;
        }
        return false;
    }
}